frontend/packages/topology/src/data-transforms/DataModelProvider.tsx

```diff
@@ -19,58 +17,104 @@ interface DataModelProviderProps {
   children?: React.ReactNode;
 }
 
-export const getNamespacedDynamicModelFactories = (
-  factories: LoadedExtension<DynamicTopologyDataModelFactory>[],
-) =>
-  factories.map(({ properties, ...ext }) => {
-    return {
-      ...ext,
-      properties: {
-        ...properties,
-        resources: (namespace: string) =>
-          Object.assign(
-            {},
-            ...Object.entries(properties.resources).map(([k, v]) => {
-              const kind = v?.model?.version
-                ? referenceForExtensionModel(v.model)
-                : v?.model
-                ? referenceForModel(modelForGroupKind(v.model?.group, v.model?.kind))
-                : v?.opts?.kind;
+const WatchResourcesHookResolver: React.FC<{
+  useResources: ExtensionHook<WatchK8sResources<any>>;
+  onResolve: (resources: WatchK8sResources<any>) => void;
+}> = ({ useResources, onResolve }) => {
+  const [resources, loaded] = useResources();
+  React.useEffect(() => {
+    if (loaded) {
+      onResolve(resources);
+    }
+  }, [loaded, onResolve, resources]);
+  return null;
+};
+
+export const DynamicWatchResourcesLoader: React.FC<{
+  children: (props: WatchK8sResources<any>) => React.ReactNode;
+}> = ({ children }) => {
+  const [watchResources, setWatchResources] = React.useState<WatchK8sResources<any>>({});
+  const [modelFactories, loaded] = useResolvedExtensions<DynamicTopologyDataModelFactory>(
+    isDynamicTopologyDataModelFactory,
+  );
+  const modelResources = React.useMemo(
+    () =>
+      modelFactories
+        ?.map((m) => ({ id: m.properties.id, resourceHook: m.properties.resources }))
+        .filter(({ resourceHook }) => !!resourceHook),
+    [modelFactories],
+  );
+
+  const onResolve = React.useCallback(
+    (resources: WatchK8sResources<any>) =>
+      setWatchResources((prevResources) => ({ ...prevResources, ...resources })),
+    [],
+  );
+
+  return (
+    <>
+      <>
+        {loaded &&
+          modelResources.map(({ id, resourceHook }) => (
+            <WatchResourcesHookResolver
+              key={id}
+              useResources={resourceHook}
+              onResolve={onResolve}
+            />
+          ))}
+      </>
+      {children(watchResources)}
```

Comment:
You need to maintain a loading state here and pass the `loaded` status to the children.